controller/lib/hal/hal_stm32.cpp

```diff
@@ -0,0 +1,856 @@
+// are individually selectable and for the most part disabled on startup.
+// Clocks to the specific peripherials need to be enabled through the 
+// RCC (Reset and Clock Controller) module before the peripherial can be
+// used.
+// Pass in the base address of the peripherial to enable it's clock
+static void EnableClock( uint32_t base )
+{
+   // I don't include all the peripherials here, just the ones
+   // that we currently use or seem likely to be used in the 
+   // future.  To add more peripherials, just look up the appropriate
+   // bit in the reference manual RCC chapter.
+   //
+   // This big case statement finds the index of the register in the 
+   // array of clock enable registers, and the bit number used to enable
+   // the clock for the specified peripherial.
+   int ndx = -1;
+   int bit = 0;
+   switch( base )
+   {
+      case DMA1_BASE:    ndx = 0; bit =  0; break;
+      case DMA2_BASE:    ndx = 0; bit =  1; break;
+      case FLASH_BASE:   ndx = 0; bit =  8; break;
+
+      case GPIO_A_BASE:  ndx = 1; bit =  0; break;
+      case GPIO_B_BASE:  ndx = 1; bit =  1; break;
+      case GPIO_C_BASE:  ndx = 1; bit =  2; break;
+      case GPIO_D_BASE:  ndx = 1; bit =  3; break;
+      case GPIO_E_BASE:  ndx = 1; bit =  4; break;
+      case GPIO_H_BASE:  ndx = 1; bit =  7; break;
+      case ADC_BASE:     ndx = 1; bit = 13; break;
+
+      case TIMER2_BASE:  ndx = 4; bit =  0; break;
+      case TIMER3_BASE:  ndx = 4; bit =  1; break;
+      case TIMER6_BASE:  ndx = 4; bit =  4; break;
+      case TIMER7_BASE:  ndx = 4; bit =  7; break;
+      case SPI2_BASE:    ndx = 4; bit = 14; break;
+      case SPI3_BASE:    ndx = 4; bit = 15; break;
+      case UART2_BASE:   ndx = 4; bit = 17; break;
+      case UART3_BASE:   ndx = 4; bit = 18; break;
+      case UART4_BASE:   ndx = 4; bit = 19; break;
+      case I2C1_BASE:    ndx = 4; bit = 21; break;
+      case I2C2_BASE:    ndx = 4; bit = 22; break;
+      case I2C3_BASE:    ndx = 4; bit = 23; break;
+
+      case I2C4_BASE:    ndx = 5; bit =  1; break;
+
+      case TIMER1_BASE:  ndx = 6; bit = 11; break;
+      case SPI1_BASE:    ndx = 6; bit = 12; break;
+      case UART1_BASE:   ndx = 6; bit = 13; break;
+      case TIMER15_BASE: ndx = 6; bit = 16; break;
+      case TIMER16_BASE: ndx = 6; bit = 17; break;
+   }
+
+   // Not sure what to do if the input address isn't found.
+   // For now I just return.  Maybe I should force a fault?
+   if( ndx < 0 ) return;
+
+   // Enable the clock of the requested peripherial
+   RCC_Regs *rcc = (RCC_Regs *)RCC_BASE;
+   rcc->periphClkEna[ndx] |= (1<<bit);
+}
+
+/******************************************************************
+ * Interrupt vector table.  The interrupt vector table is a list of
+ * pointers to the various interrupt functions.  It is stored at the
+ * very start of the flash memory.
+ *****************************************************************/
+
+// Fault handlers
+static void fault( void )
+{
+   while(1){}
```

Comment:
Does this mean that the device just freezes, or does it give any nicer indication that a fault has occurred and where?

Reply:
When a fault occurs on an ARM processor you end up here.
We can add something like a blinking LED, but in the real system we would get a watchdog timeout pretty quickly and end up with a reset.
I'm open to suggestions on better ways to handle a hard fault, but they're pretty much critical failures of the system.